### ISSUE TITLE
Support CRD's

### ISSUE THREAD

**Issue by mohnishkodnani** [arianvp/dhall-kubernetes#96]
Hi, 
I am reading up on dhall to manage our solution. We have few CRD's. Does dhall have the capability to take the CRD spec ( type definition in yaml) and convert it to a type ?
We use kubebuilder to generate the CRD's and the controller for it. I would like to have that as the source of truth. Whenever, we change the CRD type ( add/remove/update fields in the CRD) in that project, I would like the dhall type for that  CRD to reflect the changes. 
kubebuilder generates a type definition for the CRD , is it possible to take that and generate appropriate dhall type so we can use it to generate yamls to deploy those.

**Comment by Gabriella439** [comment]
@mohnishkodnani: That currently cannot be done within the Dhall language.  Typically when we bind Dhall to a particular specification we do so with a tool written in a different language.

For example, this repository creates Dhall bindings to the Kubernetes OpenAPI spec using `dhall-kubernetes-generator` tool written in Haskell

One of the reasons that the binding is not done within Dhall is that the configuration language lacks two main features that would be necessary to do so:

* Text manipulation

  `Text` literals are mostly opaque and Dhall.  In particular, it is not possible to parse or otherwise introspect on `Text` values.  At most you can concatenate or interpolate them.

  This restriction could potentially change at some point, so it's not set in stone.  For example, `Integer` values used to be similarly opaque up until recently and perhaps `Text` could go the same way in response to user demand as we slowly add additional language built-ins.

* "Full spectrum" dependent types

  Even if we could parse `Text` within Dhall we would essentially need some more sophisticated dependently-typed machinery for the language to be able to convert a `Text` value to a Dhall type.  This isn't impossible, though, but would require some sustained attention to extend the type system and builtins in order to support this.

For those two reasons it's more common to bind to specifications outside of the language

That said, it might be possible to author a tool that does what you request: generate Dhall types given a custom resource definition.  That seems like it would reasonably be within the scope of this repository to me.

I personally might not have the bandwidth to author that tool within the near future, but I could help shepherd such a contribution if somebody were interested in doing that.

**Comment by mohnishkodnani** [comment]
I wouldnt mind working on it except, i am not well versed with dhall yet and its capabilities. As I get time to ramp up I might want to do such a thing. It would be great to be able to refer to the CRD as a type in dhall. 
I think normally other clients of K8s i have seen take the swagger open API specification and generate the types in the language. 
In this case the way CRD spec atleast generated by kubebuilder framework looks like a combination of yaml and validation specifications in open API.  Ex: 
```apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: xyz.abc.com
spec:
  additionalPrinterColumns:
  - JSONPath: .spec.A
    description: A
    name: A
    type: integer
  - JSONPath: .spec.B
    description: number of B
    name: B
    type: integer
  - JSONPath: .spec.C
    description: C.
    name: C
    type: string
  - JSONPath: .spec.D
    description: D.
    name: D
    type: string
  group: abc.
  names:
    kind: xyz
    plural: xyzs
    shortNames:
    - xy
  scope: Namespaced
  subresources:
    status: {}
  validation:
    openAPIV3Schema:
      properties:
        apiVersion:
          description: 'APIVersion defines the versioned schema of this representation
            of an object. Servers should convert recognized schemas to the latest
            internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#resources'
          type: string
        kind:
          description: 'Kind is a string value representing the REST resource this
            object represents. Servers may infer this from the endpoint the client
            submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds'
          type: string
        metadata:
          type: object
        spec:
          properties:
            A:
              description: Name of the A
              type: string
            B:
              description: B
              format: int32
              type: integer
            C:
              description: enum
              enum:
              - x
              - y
              - z
              type: string
            D:
              description: D
              type: object
            E:
              description: E
              items:
                properties:
                  time:
                    type: string
                  name:
                    type: string
                  def:
                    type: string
                  ghi:
                    type: string
                  version:
                    type: string
                required:
                - def
                - ghi
                type: object
              type: array
          required:
          - A
          - B
          - C
          - D
          - E
          type: object
        status:
          properties:
            Astatus:
              format: int32
              type: integer
          type: object
  version: v1alpha1
status:
  acceptedNames:
    kind: ""
    plural: ""
  conditions: []
  storedVersions: []```

And thank you very very much for taking the time to get back to me, I really appreciate it.

**Comment by Gabriella439** [comment]
@mohnishkodnani: You're welcome! 🙂 

Yeah, this repository also generates types from the Kubernetes OpenAPI specification, but as far as I know does not yet support custom resource definitions (although I might be wrong)

**Comment by amarrella** [comment]
I have some half implemented code to handle this use case :) I'll try to open a PR or share it here

**Comment by amarrella** [comment]
I opened https://github.com/dhall-lang/dhall-kubernetes/pull/98, let me know what you think

**Comment by Gabriella439** [comment]
This is now fixed by #98
